src/Symfony/Component/DependencyInjection/Loader/FileLoader.php

```diff
@@ -56,10 +56,26 @@ public function registerClasses(Definition $prototype, $namespace, $resource, $e
 
         $classes = $this->findClasses($namespace, $resource, $exclude);
         // prepare for deep cloning
-        $prototype = serialize($prototype);
+        $serializedPrototype = serialize($prototype);
+        $interfaces = array();
+        $interfaceImplementations = array();
 
         foreach ($classes as $class) {
-            $this->setDefinition($class, unserialize($prototype));
+            if (interface_exists($class, false)) {
+                $interfaces[] = $class;
+            } else {
+                $this->setDefinition($class, unserialize($serializedPrototype));
+                foreach (class_implements($class, false) as $interface) {
+                    $interfaceImplementations[$interface][] = $class;
```

Comment:
You don't need to keep all implementations in memory. Just persist the class or set `$interfaceImplementations[$interface]` to false for example if it exists already.

Reply:
fixed